File: Linked Lists/linked_list.rb
# Implementation of a Singly Linked List

class Node
  attr_accessor :value, :next_node
  def initialize(value)
    @value = value
    @next_node = nil
    
  end
end

class LinkedList
  attr_reader :head, :tail, :length
  def initialize()
    @head = nil
    @tail = nil
    @length = 0
  end

  def add_to_tail(val)
    new_node = Node.new(val)
    
    if @head == nil
      @head = new_node
    else
      @tail.next_node = new_node
    end
    @tail = new_node
    @length += 1
    self 
  end

  def remove_tail
    return nil if @head == nil
    current = @head
    new_tail = current
    while (current.next_node)
      new_tail = current
      current = current.next_node
    end
    @tail = new_tail
    @tail.next_node = nil
    @length -= 1

    if @length == 0
      @head = nil
      @tail = nil
    end
    current 
  end

  def add_to_head(val)
    new_node = Node.new(val)
    if @head == nil
      @head = new_node
      @tail = new_node
    else
      new_node.next_node = @head
      @head = new_node
    end
    @length += 1
    self
  end

  def remove_head
    return nil if @head == nil
    current_head = @head
    @head = current_head.next_node
    @length -= 1
    if @length == 0
      @tail = nil
    end
    current_head
  end

  def contains(target)
    return nil if @head == nil
    current_node = @head
    while (current_node)
      return true if current_node.value == target
      current_node = current_node.next_node
    end
    false
  end

  def get(index)
    return nil if index < 0 || index >= @length
    counter = 0
    current = @head
    while (counter != index)
      current = current.next_node
      counter += 1
    end
    current
  end

  def set(index, val)
    found_node = self.get(index)

    if found_node
      found_node.value = val
      return true
    end
    false
  end

  def insert(index, val)
    return false if index < 0 || index >= @length 
    return !!self.add_to_tail(val) if index == @length - 1
    return !!self.add_to_head(val) if index === 0

    new_node = Node.new(val)
    prev_node = self.get(index - 1)
    temp = prev_node.next_node
    prev_node.next_node = new_node
    new_node.next_node = temp
    @length += 1
    true
  end

  def remove(index)
    return nil if index < 0 || index >= @length
    return self.remove_head if index == 0 
    return self.remove_tail if index == @length - 1

    prev_node = self.get(index - 1)
    removed = prev_node.next_node
    prev_node.next_node = removed.next_node
    @length -= 1
    removed
  end

  def size
    @length
  end
end

File: Javascript/Linked Lists/linked_list.js
// TODO: Implement a Singly Linked List Node class here
class Node {
	constructor(val) {
		this.value = val;
		this.next = null;
	}
}

// TODO: Implement a Singly Linked List class here
class LinkedList {
	constructor() {
		this.head = null;
		this.tail = null;
		this.length = 0;
	}

	// TODO: Implement the addToTail method here
	addToTail(val) {
		const newNode = new Node(val);

		if (!this.head) {
			this.head = newNode;
		} else {
			this.tail.next = newNode;
		}
		this.tail = newNode;
		this.length++;
		return this;
	}

	// TODO: Implement the removeTail method here
	removeTail() {
		if (!this.head) return undefined;
		let currNode = this.head;
		let newTail = currNode;
		while (currNode.next) {
			newTail = currNode;
			currNode = currNode.next;
		}
		this.tail = newTail;
		this.tail.next = null;
		this.length--;
		if (this.length === 0) {
			this.head = null;
			this.tail = null;
		}
		return currNode;
	}

	// TODO: Implement the addToHead method here
	addToHead(val) {
		let newNode = new Node(val);

		if (!this.head) {
			this.head = newNode;
			this.tail = newNode;
		} else {
			newNode.next = this.head;
			this.head = newNode;
		}
		this.length++;
		return this;
	}

	// TODO: Implement the removeHead method here
	removeHead() {
		if (!this.head) return undefined;

		let currHead = this.head;
		this.head = currHead.next;
		this.length--;
		if (this.length === 0) {
			this.tail = null;
		}
		return currHead;
	}

	// TODO: Implement the contains method here
	contains(target) {
		let currNode = this.head;
		while (currNode) {
			if (currNode.value === target) return true;
			currNode = currNode.next;
		}
		return false;
	}

	// TODO: Implement the get method here
	get(index) {
		if (index < 0 || index >= this.length) return null;
		let counter = 0;
		let currNode = this.head;
		while (counter !== index) {
			currNode = currNode.next;
			counter++;
		}
		return currNode;
	}

	// TODO: Implement the set method here
	set(index, val) {
		let foundNode = this.get(index);

		if (foundNode) {
			foundNode.value = val;
			return true;
		}
		return false;
	}

	// TODO: Implement the insert method here
	insert(index, val) {
		if (index < 0 || index >= this.length) return false;
		if (index === 0) return !!this.addToHead(val);
		if (index === this.length) return !!this.addToTail(val);

		const newNode = new Node(val);
		const prevNode = this.get(index - 1);
		const temp = prevNode.next;
		prevNode.next = newNode;
		newNode.next = temp;
		this.length++;
		return true;
	}

	// TODO: Implement the remove method here
	remove(index) {
		if (index < 0 || index >= this.length) return undefined;
		if (index === 0) return this.removeHead;
		if (index === this.length) return this.removeTail;

		let prev_node = this.get(index - 1);
		let removedNode = prev_node.next;
		prev_node.next = removedNode.next;
		this.length--;
		return removedNode;
	}

	// TODO: Implement the size method here
	size() {
		return this.length;
	}
}
